### PR TITLE
Fix CVE-2017-1000126

### LINKED ISSUE #175: CVE-2017-1000126: exiv2 0.26 contains a Stack out of bounds read in webp parser
In [this discussion](http://www.openwall.com/lists/oss-security/2017/06/30/1), CVE-2017-1000126 was disclosed: A malformed webp file causes a six bytes stack out of bounds read.

POC: [stack-oob-read.webp.gz](https://github.com/Exiv2/exiv2/files/1500266/stack-oob-read.webp.gz)

Stack trace:

```
==598==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffcc12aa054 at pc 0x0000004fe311 bp 0x7ffcc12a9f90 sp 0x7ffcc12a9740
READ of size 6 at 0x7ffcc12aa054 thread T0
    #0 0x4fe310 in __interceptor_memcmp.part.76 (/r/exiv2/exiv2+0x4fe310)
    #1 0x8889d0 in Exiv2::WebPImage::getHeaderOffset(unsigned char*, long, unsigned char*, long) /f/exiv2-trunk/src/webpimage.cpp:798:17
    #2 0x8889d0 in Exiv2::WebPImage::decodeChunks(unsigned long) /f/exiv2-trunk/src/webpimage.cpp:601
    #3 0x884ff2 in Exiv2::WebPImage::readMetadata() /f/exiv2-trunk/src/webpimage.cpp:496:20
    #4 0x59786a in Action::Print::printSummary() /f/exiv2-trunk/src/actions.cpp:289:16
    #5 0x596ef8 in Action::Print::run(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /f/exiv2-trunk/src/actions.cpp:244:44
    #6 0x55fb3f in main /f/exiv2-trunk/src/exiv2.cpp:170:25
    #7 0x7f7f9cc9f1d0 in __libc_start_main /var/tmp/portage/sys-libs/glibc-2.24-r2/work/glibc-2.24/csu/../csu/libc-start.c:289
    #8 0x468979 in _start (/r/exiv2/exiv2+0x468979)

Address 0x7ffcc12aa054 is located in stack of thread T0 at offset 180 in frame
    #0 0x885a0f in Exiv2::WebPImage::decodeChunks(unsigned long) /f/exiv2-trunk/src/webpimage.cpp:501

  This frame has 13 object(s):
    [32, 36) 'size_buff' (line 503)
    [48, 64) 'payload' (line 516)
    [80, 84) 'size_buf' (line 520)
    [96, 100) 'size_buf48' (line 536)
    [112, 114) 'size_buf_w' (line 551)
    [128, 131) 'size_buf_h' (line 552)
    [144, 148) 'size_buf112' (line 568)
    [160, 162) 'size_buff152' (line 587)
    [176, 180) 'exifLongHeader295' (line 588) <== Memory access at offset 180 overflows this variable
    [192, 196) 'exifTiffBEHeader297' (line 591)
    [208, 272) 'xmpData' (line 650)
    [304, 688) 'temp.lvalue'
    [752, 1136) 'temp.lvalue232'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
```

### PR DESCRIPTION
**Pull request by D4N**
This PR mainly adds a new internal helper: `enforce()` which I have stolen from the D programming language.

You can think of `enforce()` as `assert()` only that the latter should be used in places where one can be certain that the condition is always true. `enforce()` on the other hand forces the condition to be true, or it throws an exception. This is intended to be used for parameter verification, so that we can skip the tedious:
``` C++
if (param >= file_size) {
   throw Exiv2::Error(Exiv2::kerCorruptedMetadata);
}
```
and replace it with:
``` C++
EXIV2_ENFORCE(param < file_size);
```

I have used enforce to fix #175 aka CVE-2017-1000126.

### CODE REVIEW

**Comment by D4N** [comment]
Interesting, Appveyor fails, because no error occurs there. That is really odd.

**Review by clanmills — APPROVED** [review]
This looks really good to me, Dan.  Good idea. 

Something to consider:  It's ingenious that you pass/report the condition that's being enforced.  However that doesn't always to tell you what the condition means.  So we see kerEnforceFailed instead of kerCorruptedMetadata.  What do think about a second argument for EXIV2_ENFORCE(condition, errorCode)?  And the report says "enforce failed: <condition>, error: errorCode".

**Comment by clanmills** [comment]
@D4N: Something I forgot to mention.  Can you add enforce.cpp to **_src/Makefile_** please, then the build autotools (./configure) build isn't broken.  The **_msvc/_** build environment is currently broken due to renaming about 20 source files such as src/canonmn.cpp to src/canonmn_int.cpp.  I'll deal with **_msvc/_** once I've mastered the PR process.

**Comment by D4N** [comment]
I am not sold on the way `enforce()` reports errors and indeed the condition is not always quite helpful.

The problem with explicitly setting an error is, that `enforce()` just throws an exception, so if you want to set a different error code than `kerEnforceFailed` it would be more idiomatic to simply let enforce throw an `Exiv2::Error` with that error code, which can be done via the functions and not the macro:
``` C++
inline void enforce(bool condition, Exiv2::ErrorCode err_code);
template <typename T>
inline void enforce(bool condition, Exiv2::ErrorCode err_code, const T& arg1);
// and so on, these are in src/enforce.hpp
```
You won't get the error message `enforce failed: ...` error message though but the standard message instead. We could also completely remove `kerEnforceFailed` and require everyone to throw meaningful exceptions, which when I now think about it, might be the better way.

Concerning the autotools build: it shouldn't be changed by this PR, enforce is header only, so there should be no need to add it into build lists. Unless the autotools build also compiles the unit tests.

However, before this gets merged, I have to find out why on Earth there is no error on Windows.

**Comment by clanmills** [comment]
@D4N As always, you put forward compelling explanations.  I'm suggesting that errorCode is reported as a (localised) string, however the code should proceed to throw the enforce exception.

My mistake about enforce.hpp which I mistakenly thought was enforce.cpp.  So the autotools build should be OK.

I looked at the Windows/AppVeyor build log and without more knowledge of the python test harness, I didn't know what was wrong.

File: src/webpimage.cpp [hunk 1]
```diff
@@ -508,7 +509,8 @@ namespace Exiv2 {
         while ( !io_->eof() && (uint64_t) io_->tell() < filesize) {
             io_->read(chunkId.pData_, WEBP_TAG_SIZE);
             io_->read(size_buff, WEBP_TAG_SIZE);
-            long size = Exiv2::getULong(size_buff, littleEndian);
+            const long size = Exiv2::getULong(size_buff, littleEndian);
+            EXIV2_ENFORCE(size < filesize - io_->tell());
```
Comment (piponazo) [inline src/webpimage.cpp]: Tiny suggestion. I would write this code as:

```
EXIV2_ENFORCE(size < (filesize - io_->tell()));
```
It's always clearer for me being explicit about the operators than relying in the operator precedence rules.

**Review by piponazo — APPROVED** [review]
Nice contribution!

**Comment by piponazo** [comment]
@D4N Do you have a windows development environment where to test why the test fails? If not, I could try to give it a try on my Windows machine.

**Comment by clanmills** [comment]
I'm happy to do the build on Cygwin and CMake/msvc on Windows 10.

You can access the Windows 10 VM on the build server in two steps:
```
$ ssh yourusername@eviv2.dyndns.org
.... you're in MacOS-X on the buildserver ....
$ ssh rmills-w7
.... you're on in a Cygwin bash shell on the Windows 10 VM .....
(for historical reasons it's called rmillsmm-w7)
```

**Comment by D4N** [comment]
@piponazo I don't have a Windows development machine and even if I had, I have never used Windows for anything serious, so I have my doubts that I could achieve anything with one (except installing Linux on it ;-) ). I'll give Robin's VM a try though.

@clanmills I have some trouble accessing the Windows VM, I only get a:
```
rmillsmm:~ dan$ ssh rmills-w7
ssh: Could not resolve hostname rmills-w7: nodename nor servname provided, or not known
```
Do you have the IP of the VM? And another stupid question: how do I debug something on windows via the CLI?

**Comment by clanmills** [comment]
The IP of rmillsmm-w7 is 192.168.0.201 and is in /etc/hosts.

If you build in the Cygwin bash shell using cmake, you'll be able to debug with gdb.  Cygwin is a good port of POSIX.  You should feel very comfortable.

You can build with CMake/msvc (look in the appveyor config file).  I've not idea how to debug from the command-line with those tools.  Perhaps a few well placed printf's will be sufficient.

I'm very happy to use Visual Studio.  I'll build and debug it for you if you brief me about what you think could be wrong.

**Comment by clanmills** [comment]
My typo guys.  The W10 VM is **rmillsmm-w7** (and not the rmills-w7 typo).

**Comment by clanmills** [comment]
I’ll have to investigate in the morning.  I thought the Cygwin ssh server has no security and will log you on without a username or password!

Let’s get to bed.  Tomorrow’s another day.

Robin
http://clanmills.com
Not sent from my iPhone

**Comment by D4N** [comment]
Good news, this is not a windows specific issue. It is caused by different build options between Travis/me and Appveyor. I just tried building exiv2 in the same way as Appveyor does and I can reproduce the test failure. I'll take a look at this later today.

@piponazo What's your opinion on `enforce()`?

**Comment by piponazo** [comment]
Definitely better than `assert()` so we can place it in production code and add tests dealing with that code (something that we cannot do with `assert()` if you want to run tests in Debug mode) ;)

**Comment by D4N** [comment]
Yesterday I investigated the issue further and it turned out, that the different build options only made the underlying issue visible: I was fixing the issue in the wrong function! I assumed that the problem was happening in `WebPImage::decodeChunks(filesize)`, and that the `filesize` parameter was correct. Well, turns out, it wasn't. It is extracted in `WebPImage::readMetadata()` from the input file and not checked. The reproducer is so small, that during reading of the `filesize` it encounters eof and only the first bytes of the parameter are actually valid. The last ones are random garbage, so the actual contents depend on the compiler, OS version & compiler flags. On travis the random garbage resulted in a filesize that triggered the `enforce` that I added, but on appveyor that was not the case.

Long story short: I have added another `enforce` into `WebPImage::readMetadata()` that the extracted filesize is actually less or equal to the actual size of the image. Now the new test suite passes, but the old test suite fails. Great -.-

I'll split this PR into two and create a new one for `enforce()`, so that we can get that one merged. @clanmills Do you happen to know the webp standard?

**Comment by clanmills** [comment]
@D4N Thanks for working on this.  I know enough about WebP to deal with this and to fix the test harness.  The WebP code was contributed by Ben and I worked with him to add WebPImage::writeMetadata() and test/webp.sh

**Comment by D4N** [comment]
@clanmills Thanks a lot! Basically my problem is the following in `src/webpimage.cpp:493` we call `WebPImage::decodeChunks()` and pass `Exiv2::getULong(data + WEBP_TAG_SIZE, littleEndian) + 12` as the `filesize` parameter. I have added a check that this parameter is less or equal to the actual size of the file, which now makes the test suite fail.

Do you happen to know whether this parameter is actually the size of the file? If that is the case, then we should maybe just ignore it and pass the actual size of the file into `WebPImage::decodeChunks()` instead?

**Comment by clanmills** [comment]
It's been a while and my old braincell doesn't remember too well.  The spec's here:  https://developers.google.com/speed/webp/docs/riff_container  It's reading this:
```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                         Chunk FourCC                          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                          Chunk Size                           |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                         Chunk Payload                         |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

I think there's a series/sequence of [header/payload]+

The code isn't cautious and naively assumes the data will be good.  The fuzzing files are exploiting this.  So:
```
            io_->read(chunkId.pData_, WEBP_TAG_SIZE);
            io_->read(size_buff, WEBP_TAG_SIZE);
            long size = Exiv2::getULong(size_buff, littleEndian);

            DataBuf payload(size);
```

We need to test that the reads did read 4 bytes:
```
767 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $ find . -name "*pp" -exec grep -H --line-number WEBP_TAG_SIZE {} \; | grep x4
./src/webpimage.cpp:79:    const int WebPImage::WEBP_TAG_SIZE = 0x4;
768 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $
``` 
`long size` must be less than io_->size().  More accurately:

```
checkOverflow(io_->tell()+size) < io_->size()
```

I don't know if a function checkOverflow exists to use your beautiful safe code.  These fuzzing files are very slippery.  They can exploit the code in many ways.  We have to be very very cautious. 

You've said that the test suite fails.  test/webp-test.sh?   Or another test?

**Comment by D4N** [comment]
I think I have fixed the problem (the CI will still be red though, as the enforce PR is not merged in yet).

The issue is two fold:
1. The webp parser had a tiny bug. The function `WebPImage::decodeChunks` was called with the extracted file size plus 12. This is wrong, it should be plus 8, as the document that @clanmills linked states.
2. My enforce call was checking `size < (filesize - io_->tell())` which will fail when the current chunk ends at the last byte of the file. Now it checks for equality too.

The testsuite works for me locally when I pull in `enforce.hpp` from #252, so once that is merged, this should be green too.

**Pull request merged**